nri_device_injector/nri_device_injector.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2023 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"golang.org/x/sys/unix"
+	"sigs.k8s.io/yaml"
+
+	"github.com/containerd/nri/pkg/api"
+	"github.com/containerd/nri/pkg/stub"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	deviceKeyPrefix = "devices.gke.io"
+	// Key prefix for device injection to a container, followed by container name
+	ctrDeviceKeyPrefix = deviceKeyPrefix + "/container."
+	pluginName         = "device_injector_nri"
+	pluginIdx          = "10"
+	// Device types.
+	blockDevice = "b"
+	charDevice  = "c"
+	fifoDevice  = "p"
+)
+
+type device struct {
+	Path     string `json:"path"`
+	Type     string `json:"type"`
+	Major    int64  `json:"major"`
+	Minor    int64  `json:"minor"`
+	FileMode uint32 `json:"file_mode"`
+	UID      uint32 `json:"uid"`
+	GID      uint32 `json:"gid"`
+}
+
+type plugin struct {
+	stub stub.Stub
+}
+
+func main() {
+	var (
+		opts []stub.Option
+		err  error
+	)
+
+	opts = append(opts, stub.WithPluginName(pluginName))
+	opts = append(opts, stub.WithPluginIdx(pluginIdx))
+
+	p := &plugin{}
+
+	if p.stub, err = stub.New(p, append(opts, stub.WithOnClose(p.onClose))...); err != nil {
+		log.Errorf("Failed to create plugin stub: %v", err)
+		os.Exit(1)
+	}
+
+	err = p.stub.Run(context.Background())
+	if err != nil {
+		log.Errorf("plugin exited with error %v", err)
+		os.Exit(1)
+	}
+}
+
+func (p *plugin) onClose() {
+	log.Info("NRI connection closed")
+}
+
+// CreateContainer handles CreateContainer requests relayed to the plugin by containerd NRI.
+// The plugin makes adjustment on containers with device injection annotations.
+func (p *plugin) CreateContainer(_ context.Context, pod *api.PodSandbox, container *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error) {
+	var (
+		ctrName string
+		devices []device
+		err     error
+	)
+
+	ctrName = container.Name
+	log.Infof("Started CreateContainer on container: %v, pod: %v, namespace: %v", ctrName, pod.Name, pod.Namespace)
```

Comment:
```suggestion
	l := log.WithField("container", ctrName).WithField("pod", pod.Name).WithField("namespace", pod.Namespace)
	l.Info("Started CreateContainer")
```

We can then reuse `l` to keep this context throughout the rest of the function.

(If we end up wanting to pass it around elsewhere, the pattern used inside containerd is `github.com/containerd/log` which associates a logger with a `context.Context` and then accesses with `log.G(ctx)...`.)

Reply:
Done, thanks!